### PR TITLE
Fix the string search behavior when using ICU

### LINKED ISSUE #43736: Breaking change with string.IndexOf(string) from .NET Core 3.0 -> .NET 5.0
### Description

I'm extending a package to support .NET 5.0 and ran into a breaking change. Given the console application:

```csharp
using System;

namespace ConsoleApp
{
    class Program
    {
        static void Main(string[] args)
        {
            var actual = "Detail of supported commands\n============\n## Documentation produced for DelegateDecompiler, version 0.28.0 on Thursday, 22 October 2020 16:03\n\r\nThis file documents what linq commands **DelegateDecompiler** supports when\r\nworking with [Entity Framework Core](https://docs.microsoft.com/en-us/ef/core/) (EF).\r\nEF has one of the best implementations for converting Linq `IQueryable<>` commands into database\r\naccess commands, in EF's case T-SQL. Therefore it is a good candidate for using in our tests.\r\n\r\nThis documentation was produced by compaired direct EF Linq queries against the same query implemented\r\nas a DelegateDecompiler's `Computed` properties. This produces a Supported/Not Supported flag\r\non each command type tested. Tests are groups and ordered to try and make finding things\r\neasier.\r\n\r\nSo, if you want to use DelegateDecompiler and are not sure whether the linq command\r\nyou want to use will work then clone this project and write your own tests.\r\n(See [How to add a test](HowToAddMoreTests.md) documentation on how to do this). \r\nIf there is a problem then please fork the repository and add your own tests. \r\nThat will make it much easier to diagnose your issue.\r\n\r\n*Note: The test suite has only recently been set up and has only a handful of tests at the moment.\r\nMore will appear as we move forward.*\r\n\r\n\r\n### Group: Unit Test Group\n#### [My Unit Test1](../TestGroup01UnitTestGroup/Test01MyUnitTest1):\n- Supported\n  * Good1 (line 1)\n  * Good2 (line 2)\n\r\n#### [My Unit Test2](../TestGroup01UnitTestGroup/Test01MyUnitTest2):\n- Supported\n  * Good1 (line 1)\n  * Good2 (line 2)\n\r\n\r\n\nThe End\n";

            var expected = "\n#### [My Unit Test2](";

            Console.WriteLine($"actual.Contains(expected): {actual.Contains(expected)}");
            Console.WriteLine($"actual.IndexOf(expected): {actual.IndexOf(expected)}");
        }
    }
}
```

I get different results based on the runtime from .NET Core 3.0 -> .NET 5.0:

.NET Core 3.0:

```
actual.Contains(expected): True
actual.IndexOf(expected): 1475
```

.NET 5.0:

```
actual.Contains(expected): True
actual.IndexOf(expected): -1
```

### Configuration

Windows 10 Pro Build 19041 x64
.NET Core 3.1.9
.NET 5.0.0-rc.2.20475.5

### Regression?

Yes, it worked through .NET Core 3.1.9

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes https://github.com/dotnet/runtime/issues/43736

In .NET 5.0, we have switched to use the ICU library for Globalization support. We got some issues with the string search operations like `IndexOf, LastIndexOf, IsSuffix, and IsPrefix` APIs. The linked issues containing some details about such problems. One example, when searching for "\n" in the string "\r\n" get string not found as a result. 
The underlying cause of this behavior is ICU internally is using a break iterator which disallows breaking between some characters.
The provided solution here is we are customizing the break iterator to get the desired behavior.

### CODE REVIEW

**Comment by tarekgh** [comment]
@safern could you please have a look? 

CC @danmoseley @ericstj @stephentoub @GrabYourPitchforks

**Comment by GrabYourPitchforks** [comment]
Haven't looked at the code yet, but this seems like an _extremely_ risky change to make right before an RC snap. Are we sure we want this for 6.0?

**Comment by tarekgh** [comment]
> Haven't looked at the code yet, but this seems like an extremely risky change to make right before an RC snap. Are we sure we want this for 6.0?

Why you think it is that risky? I am not seeing the same. Even I could think can meet the serving bar if we needed to service 5.0.

File: src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c
Comment (GrabYourPitchforks) [inline src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:464]: What about other rules like GB6 - GB8, which control Hangul searching; or GB12 - GB13, which control regional indicators?

For example, let's say that I have the string `"\U0001F1F5\U0001F1F7\U0001F1FA\U0001F1F8"`. This is the regional indicator for the Puerto Rican flag followed by the regional indicator for the USA flag. If I search for the substring `"\U0001F1F7\U0001F1FA"` (regional indicator for the Russian flag), will this result in a match? The normal break rules ensure that this will not result in a match. Should we have a unit test for this to ensure that we didn't accidentally nuke these break rules?

File: src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c
Comment (tarekgh) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:471]: The whole BreakIteratorRule is defined in ASCII range. So, we shouldn't be affected by the codepage settings at all.

File: src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c
Comment (tarekgh) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:464]: I didn't include such rules for two reasons: 
- It is different than `NLS` behavior. Users usually complain about the behavior change from `NLS.
- The most important reason is we need to differentiate between the string search operation and unit/word/phrase breaking. These are really two different operations and shouldn't be handled differently. If I am parsing a document and trying to find the boundaries of the characters, words, or phrases, then these rules make sense to apply at that time. But if I am searching for a string, these rules don't make sense at that time. Think about as `\r\n` case. If trying to find the boundaries, then not allow breaking between `\r` and `\n` make sense. But if you search `\n` it doesn't make sense not finding it in `\r\n`. Also, ICU has a bug that by default using the break iterator while the documentation is clear telling if not intentionally passing a breaker, then shouldn't break between characters and only the collation operation only should determine the search result. 

in https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/usearch_8h.html#ac0bf8411b693a14ed35d2a8883225cd7 it states:

```
breakiter | A BreakIterator that will be used to restrict the points at which matches are detected. If a match is found, but the match's start or end index is not a boundary as determined by the BreakIterator, the match will be rejected and another will be searched for. If this parameter is NULL, no break detection is attempted
```

Comment (tarekgh) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:464]: I want to add, I prefer not adding any more breaking rules from now till we get a real request users want it. Later we can allow our API users to customize the breaking behavior as they want by providing some options doing that.

**Comment by tarekgh** [comment]
@GrabYourPitchforks just checking if you have any more comments? 

@safern could you please have a look when you have a chance?

**Review by GrabYourPitchforks — APPROVED** [review]

**Comment by GrabYourPitchforks** [comment]
I don't have any other feedback. I still maintain that this change has the potential to introduce subtle breaks since we're introducing NLS-like behaviors across all platforms, even non-Windows platforms which always used ICU under the covers. And that normally we'd want such a change to bake for several previews before it goes in for real. I don't feel confident that we understand the potential fallout well enough to introduce this in RC, let alone in-place servicing for downlevel platforms.

All of that said, if we still believe it's the right change for the product, I'll accept it. The code looks solid enough. :)

**Comment by danmoseley** [comment]
@stephentoub @marklio what are your thoughts on the breakingness here.

**Comment by danmoseley** [comment]
I'd like to get more perspectives before merging, if thats OK?

**Comment by GrabYourPitchforks** [comment]
@danmoseley Check your inbox. Tarek sent an excellent email that laid out the problem and why this solution was chosen. My response to that email boiled down to "let's make sure we understand these three things first, and if we're good with them then hit GO!"

**Comment by danmoseley** [comment]
Marking while we cross i's and dot t's.

**Comment by tarekgh** [comment]
@aolszowka you have reacted to a couple of comments here. do you have any feedback we need to consider with this change? or you are just watching the progress?

File: src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c
Comment (GrabYourPitchforks) [inline src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:543]: Nit: for array allocation, best practice is to prefer `calloc((size_t)breakIteratorRulesLength + 1, sizeof(UChar))` over `malloc(numElements * sizeOfT)`. I wouldn't hold up the PR for this, but if you do a quality pass in the future we should consider changing this call site.

File: src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c
Comment (GrabYourPitchforks) [inline src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:549]: SDL: The `strcpy` family of functions is banned and must not be used in new code. (This new call site doesn't qualify for grandfathering.) Use [`u_uastrncpy`](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/ustring_8h.html#a54116c3e615b863b5857c255568ffab9) instead, which is in the approved `strncpy` family of functions.
(Same comment line 555.)

File: src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c
Comment (GrabYourPitchforks) [inline src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:531]: This access requires acquire semantics. Otherwise the read from *s_breakIteratorRulesLength* can be reordered to occur before the read from *s_breakIteratorRules*, which sets up a race condition:

1. Assume both statics have their default value of nullptr / 0.
2. Thread A reads *s_breakIteratorRulesLength* (value = 0).
3. Thread A is interrupted.
4. Thread B runs through this method and sets both *s_breakIteratorRulesLength* and *s_breakIteratorRules*. (The latter store uses release semantics because of the CAS operation.)
5. Thread A is resumed.
6. Thread A reads a _non-null_ value for *s_breakIteratorRules*.
7. Thread A would have asserted on line 533, but since asserts are compiled out in release builds, execution progresses to line 534 and results in incorrect behavior.

Something that may be easier is to drop the *s_breakIteratorRulesLength* static entirely and to call instead `ubrk_openRules(s_breakIteratorRules, -1, ...)`. That way this code never has to worry about accesses to these two statics being reordered since there are no longer two statics to worry about.

**Review by GrabYourPitchforks — CHANGES_REQUESTED** [review]

**Comment by aolszowka** [comment]
@tarekgh 

> @aolszowka you have reacted to a couple of comments here. do you have any feedback we need to consider with this change? or you are just watching the progress?

I am just watching this progress. I reported one of the original repros on the ICU behaviors here: https://github.com/dotnet/runtime/issues/43802. I no longer work for that particular company with that code base that would have been greatly affected.

I can say that these types of low level changes often result in extremely high profile breaks that doesn't generate a lot of good will at the C-Level, making the case for staying current on .NET difficult. From the community response in the various threads I don't think we're alone in that situation.

I would echo @GrabYourPitchforks comments, and would be interested to see what persuaded a change to say 'ship it':

>I don't have any other feedback. I still maintain that this change has the potential to introduce subtle breaks since we're introducing NLS-like behaviors across all platforms, even non-Windows platforms which always used ICU under the covers. And that normally we'd want such a change to bake for several previews before it goes in for real. I don't feel confident that we understand the potential fallout well enough to introduce this in RC, let alone in-place servicing for downlevel platforms.

The feedback I have given to at least alert developers (at compile time) of these types of gotchas seems to have been back burner-ed https://github.com/dotnet/runtime/issues/47775. In my opinion, in a perfect world these types of changes would roll out first as a compiler warning for a few months to encourage devs to study potential problem areas for remediation (or at least submit knowledge bases for areas to look out for) and then perform a runtime roll out. However historically these Analyzers have been an after thought, only after some hard-won in the wild knowledge has cause incidents.

File: src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c
Comment (tarekgh) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:543]: Why need to force the initialization if the content would be overwritten anyway? perf wise malloc is better in this situation.

Comment (GrabYourPitchforks) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:543]: It's not a requirement, just a preference. It avoids the problems of: (a) integer overflow due to the multiplication; and (b) accidentally using uninitialized memory. Neither of these issues affects this call site, but I've seen some chatter about new rules related to malloc / calloc being introduced in the future. We could choose to proactively address this call site now, or we could wait until those rules are formalized and then do a pass over the entire code base in the future.

File: src/libraries/Native/Unix/System.Globalization.Native/pal_timeZoneInfo.c
Comment (GrabYourPitchforks) [inline src/libraries/Native/Unix/System.Globalization.Native/pal_timeZoneInfo.c:227]: This should be `TZID_LENGTH`, not `(int32_t)strlen(testId)`. (It's the size of the destination, not the size of the source.)

**Review by GrabYourPitchforks — APPROVED** [review]

File: src/libraries/Native/Unix/System.Globalization.Native/pal_timeZoneInfo.c
Comment (tarekgh) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_timeZoneInfo.c]: That make sense.

**Comment by tarekgh** [comment]
The failure in the CI leg `runtime (Libraries Test Run checked coreclr windows x86 Release) ` is unrelated to the change and tracked by https://github.com/dotnet/runtime/issues/57452

**Pull request merged**
